docs/design/jet/023-jobupload.md

```diff
@@ -0,0 +1,147 @@
+
+#### Client Related Changes
+A new method has been added to JetService interface. Any client that implements this interface, has to implement this method
+```java
+void uploadJob(@Nonnull Path jarPath, String snapshotName, String jobName, String mainClass, List<String> jobParameters);
+```
+
+### Technical Design
+
+The client protocol needs to support job uploading. So that non-java clients can also upload jet jobs.
+For this purpose two new messages are added to client protocol
+
+1. uploadJobMetaData
+2. uploadJobMultipart
+
+**1. uploadJobMetaData Message**
+
+The upload process starts with uploadJobMetaData. This message contains
+
+| Term                   | Definition                                                       |
+|------------------------|------------------------------------------------------------------|
+| sessionId | The UUID. This field associates all messages in a session |
+| jarSize | The jar size in bytes                                            |
+| snapshotName | Argument passed when starting the job                            |
+| jobName | Argument passed when starting the job                            |
+| mainClass | Argument passed when starting the job                            |
+| jobParameters | Argument passed when starting the job                            |
+
+
+Upon reception of uploadJobMetaData message, the server performs some validation rules. If any of them fail, a **JetException** is thrown. If all is well, server stores a new entry in the JobUploadStore class
+
+**2. uploadJobMultipart message**
+
+The upload process continues with this message. It contains jar's bytes. This message contains these fields
+
+| Term                   | Definition                                                              |
+|------------------------|-------------------------------------------------------------------------|
+| sessionId | It is explained in the previous message |
+| currentPartNumber | It starts from 1 and shows the sequence number of the part. For example 1 of 5 |
+| totalPartNumber | It is the total number of parts of the sequence  |
+| partData | is the **byte[]** containing jar data  |
+| partSize | shows how many bytes of the partData byte[] is valid, |
+
+**Why do we need an extra partSize field?** 
+
+For optimization, it is assumed that partData is allocated only once. So we need another field to indicate the number of bytes to be read in this buffer 
+
+Upon reception of uploadJobMultipart, various checks are performed on the message and current session. If any of them fail, a **JetException** is thrown. If all is well, message is processed.
+If it is the first message a new temporary file is created. Then the partSize bytes of partData byte[] is appended to this file
+When all the parts are complete, a new job is started using HazelcastBootstrap.executeJar() call. When the job starts, the jar is loaded into memory and it is not possible to delete it anymore, because it is in use.
```

Comment:
Will the server keep track of the received parts, so if a client sends the same part again, will it handle that (maybe by discarding the duplicates)?

Reply:
Added this
Validate currentPart + 1 is not different from receivedCurrentPart - A message from the future. This also means that a duplicate message will be rejected and upload operation will fail.